machine-runner/src/design/event.ts

```diff
@@ -61,20 +62,48 @@ export namespace MachineEvent {
    * ])
    */
   export const design = <Key extends string>(key: Key): EventFactoryIntermediate<Key> => ({
+    withZod: (zodDefinition) => ({
+      [FactoryInternalsAccessor]: {
+        zodDefinition: zodDefinition,
+      },
+      type: key,
+      make: (payload) => ({
+        ...zodDefinition.parse(payload),
+        type: key,
+      }),
+    }),
+
     withPayload: () => ({
+      [FactoryInternalsAccessor]: {
+        zodDefinition: undefined,
+      },
       type: key,
       make: (payload) => ({
         ...payload,
         type: key,
       }),
     }),
+
     withoutPayload: () => ({
+      [FactoryInternalsAccessor]: {
+        zodDefinition: undefined,
+      },
       type: key,
       make: () => ({ type: key }),
     }),
   })
 
   type EventFactoryIntermediate<Key extends string> = {
+    /**
+     * Declares the payload type for this MachineEvent using zod definition.
+     *
+     * Event payload will be serialized. Payload definition cannot have fields
+     * that cannot be serialized and deserialized back via JSON.stringify and
+     * JSON.parse.
+     */
+    withZod: <Payload extends utils.SerializableObject>(
```

Comment:
could this not be unified with `withPayload<Payload>(z?: z.ZodType<Payload>)`?

Reply:
It could, but:
- different method may make a good global code search heuristic (e.g. searching the whole codebase for events that isn't zod-reinforced yet, in the case of refactoring)
- different method prevent us from being locked with zod (e.g. in case a powerful user wants io-ts support instead of zod)

Reply:
Good points, agreed!